src/Symfony/Component/RateLimiter/ClockTrait.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\RateLimiter;
+
+use Psr\Clock\ClockInterface;
+
+/**
+ * @internal
+ */
+trait ClockTrait
+{
+    private ?ClockInterface $clock;
+
+    public function setClock(?ClockInterface $clock): void
```

Comment:
does this need to be public in classes using this trait ? And if it does, does it need to be non-internal ?

Reply:
It's public because we need to call it after fetching states from the storages. But for that we don't need to have it non-internal. Though I was under the impression that the method is internal as the trait is internal. Isn't that how it works?

Reply:
As using a trait copies its method to the class (a trait is compiler-assisted copy/paste), I don't think IDEs and static analyzers would report it as internal (as they would see that as a usage of `SlidingWindow::setClock`). And if the class using the trait is not internal, its public methods must be covered by BC.

We definitely have internal traits being used to implement some public APIs of some classes, so I don't think we consider that public methods in an internal trait are automatically internal.

Reply:
makes sense, I added the `@internal` annotation to `setClock()` too